File: backend/app/services/logistics_job_service.py
from datetime import datetime
from decimal import Decimal

from sqlalchemy.orm import selectinload

from app.constants import (
    LOGISTICS_DELIVERY_SPEEDS,
    LOGISTICS_ENABLED_PAYMENT_METHODS,
    LOGISTICS_IN_HOUSE_FEE_JMD,
    LOGISTICS_IN_HOUSE_PARISHES,
    LOGISTICS_ISLAND_FEE_JMD,
    LOGISTICS_ITEM_CATEGORIES,
    LOGISTICS_ITEM_CATEGORY_LABELS,
    LOGISTICS_JOB_OPEN_STATUSES,
    LOGISTICS_PAYMENT_METHODS,
    LOGISTICS_VEHICLE_TYPES,
)
from app.extensions import db
from app.models.logistics_job import LogisticsJob
from app.models.user import User
from app.services.delivery_address_service import create_delivery_address, get_delivery_address


def _decimal(value) -> Decimal:
    return Decimal(str(value)).quantize(Decimal("0.01"))


def _load_options():
    return (
        selectinload(LogisticsJob.customer),
        selectinload(LogisticsJob.pickup_address),
        selectinload(LogisticsJob.dropoff_address),
        selectinload(LogisticsJob.in_progress_by),
        selectinload(LogisticsJob.picked_up_by),
        selectinload(LogisticsJob.in_transit_by),
        selectinload(LogisticsJob.driver_confirmed_by),
        selectinload(LogisticsJob.completed_by),
        selectinload(LogisticsJob.assigned_clerk),
        selectinload(LogisticsJob.assigned_by),
        selectinload(LogisticsJob.rejected_by),
    )


def _query():
    return LogisticsJob.query.options(*_load_options())


def _next_reference() -> str:
    prefix = "LD"
    latest = (
        LogisticsJob.query.filter(LogisticsJob.reference.like(f"{prefix}-%"))
        .order_by(LogisticsJob.requested_at.desc())
        .first()
    )
    if latest and latest.reference.startswith(f"{prefix}-"):
        try:
            seq = int(latest.reference.split("-", 1)[1]) + 1
        except ValueError:
            seq = LogisticsJob.query.count() + 1
    else:
        seq = LogisticsJob.query.count() + 1
    return f"{prefix}-{seq:05d}"


def _resolve_address(
    customer: User,
    *,
    address_id=None,
    address_data: dict | None = None,
    default_label: str,
):
    if address_id:
        address = get_delivery_address(customer, address_id)
        if not address:
            raise ValueError(f"{default_label} address not found")
        return address

    if address_data and isinstance(address_data, dict):
        payload = dict(address_data)
        if not (payload.get("label") or "").strip():
            payload["label"] = default_label
        return create_delivery_address(customer, payload)

    raise ValueError(f"{default_label} address is required")


def estimate_logistics_fee(pickup_parish: str, dropoff_parish: str) -> tuple[Decimal, bool]:
    pickup_in_house = pickup_parish in LOGISTICS_IN_HOUSE_PARISHES
    dropoff_in_house = dropoff_parish in LOGISTICS_IN_HOUSE_PARISHES
    if pickup_in_house and dropoff_in_house:
        return LOGISTICS_IN_HOUSE_FEE_JMD, False
    if pickup_in_house or dropoff_in_house:
        return LOGISTICS_ISLAND_FEE_JMD, False
    return Decimal("0.00"), True


def list_customer_logistics_jobs(customer: User, limit: int = 50) -> list[LogisticsJob]:
    return (
        _query()
        .filter_by(customer_id=customer.id)
        .order_by(LogisticsJob.requested_at.desc())
        .limit(limit)
        .all()
    )


def list_pending_logistics_jobs(limit: int = 100) -> list[LogisticsJob]:
    return (
        _query()
        .filter_by(status="pending")
        .order_by(LogisticsJob.requested_at.asc())
        .limit(limit)
        .all()
    )


def list_open_logistics_jobs(limit: int = 100) -> list[LogisticsJob]:
    return (
        _query()
        .filter(LogisticsJob.status.in_(LOGISTICS_JOB_OPEN_STATUSES))
        .order_by(LogisticsJob.requested_at.asc())
        .limit(limit)
        .all()
    )


def list_all_logistics_jobs(limit: int = 200) -> list[LogisticsJob]:
    return (
        _query()
        .order_by(LogisticsJob.requested_at.desc())
        .limit(limit)
        .all()
    )


def list_logistics_job_history(limit: int = 200) -> list[LogisticsJob]:
    return (
        _query()
        .filter(~LogisticsJob.status.in_(LOGISTICS_JOB_OPEN_STATUSES))
        .order_by(LogisticsJob.requested_at.desc())
        .limit(limit)
        .all()
    )


def list_logistics_jobs_by_status(status: str, limit: int = 200) -> list[LogisticsJob]:
    return (
        _query()
        .filter_by(status=status)
        .order_by(LogisticsJob.requested_at.desc())
        .limit(limit)
        .all()
    )


def count_open_logistics_jobs() -> int:
    return LogisticsJob.query.filter(LogisticsJob.status.in_(LOGISTICS_JOB_OPEN_STATUSES)).count()


def count_clerk_open_logistics_jobs(clerk: User) -> int:
    return (
        LogisticsJob.query.filter(
            LogisticsJob.assigned_clerk_id == clerk.id,
            LogisticsJob.status.in_(LOGISTICS_JOB_OPEN_STATUSES),
        ).count()
    )


def list_clerk_logistics_jobs(clerk: User, status: str = "active", limit: int = 100) -> list[LogisticsJob]:
    base = _query().filter(LogisticsJob.assigned_clerk_id == clerk.id)
    if status == "pending":
        jobs = base.filter_by(status="pending")
    elif status == "active":
        jobs = base.filter(LogisticsJob.status.in_(LOGISTICS_JOB_OPEN_STATUSES))
    elif status == "all":
        jobs = base
    elif status == "history":
        jobs = base.filter(~LogisticsJob.status.in_(LOGISTICS_JOB_OPEN_STATUSES))
    elif status in ("rejected", "cancelled", "completed", "picked_up", "in_transit"):
        jobs = base.filter_by(status=status)
    else:
        jobs = base.filter_by(status=status)
    return jobs.order_by(LogisticsJob.requested_at.asc()).limit(limit).all()


def count_pending_logistics_jobs() -> int:
    return LogisticsJob.query.filter_by(status="pending").count()


def get_logistics_job(job_id) -> LogisticsJob | None:
    import uuid

    try:
        jid = uuid.UUID(str(job_id))
    except ValueError:
        return None
    return _query().filter_by(id=jid).first()


def resolve_item_description(
    *,
    item_category: str | None = None,
    item_other_detail: str | None = None,
    item_description: str | None = None,
) -> str:
    if item_category:
        category = item_category.strip().lower()
        if category not in LOGISTICS_ITEM_CATEGORIES:
            raise ValueError("Select a valid item category")
        if category == "other":
            detail = (item_other_detail or "").strip()
            if not detail:
                raise ValueError("Describe the item when selecting Other")
            if len(detail) > 480:
                raise ValueError("Item description must be 500 characters or fewer")
            return f"Other — {detail}"
        return LOGISTICS_ITEM_CATEGORY_LABELS[category]

    description = (item_description or "").strip()
    if not description:
        raise ValueError("Item category is required")
    if len(description) > 500:
        raise ValueError("Item description must be 500 characters or fewer")
    return description


def resolve_vehicle_type(vehicle_type: str | None) -> str:
    value = (vehicle_type or "").strip().lower()
    if value not in LOGISTICS_VEHICLE_TYPES:
        raise ValueError("Select a vehicle size for this delivery")
    return value


def resolve_delivery_speed(delivery_speed: str | None) -> str:
    value = (delivery_speed or "immediate").strip().lower()
    if value not in LOGISTICS_DELIVERY_SPEEDS:
        raise ValueError("Select when you need this delivery")
    return value


def resolve_payment_method(payment_method: str | None) -> str:
    value = (payment_method or "cash").strip().lower()
    if value not in LOGISTICS_PAYMENT_METHODS:
        raise ValueError("Select a payment method")
    if value not in LOGISTICS_ENABLED_PAYMENT_METHODS:
        raise ValueError("Online payment is not available yet — please choose cash")
    return value


def create_logistics_job(
    customer: User,
    *,
    pickup_address_id=None,
    pickup_address: dict | None = None,
    dropoff_address_id=None,
    dropoff_address: dict | None = None,
    item_category: str | None = None,
    item_other_detail: str | None = None,
    item_description: str | None = None,
    vehicle_type: str | None = None,
    delivery_speed: str | None = None,
    payment_method: str | None = None,
    weight_lbs=None,
    notes: str | None = None,
) -> LogisticsJob:
    description = resolve_item_description(
        item_category=item_category,
        item_other_detail=item_other_detail,
        item_description=item_description,
    )
    vehicle = resolve_vehicle_type(vehicle_type)
    speed = resolve_delivery_speed(delivery_speed)
    pay_method = resolve_payment_method(payment_method)

    note_text = (notes or "").strip() or None
    if note_text and len(note_text) > 500:
        raise ValueError("notes must be 500 characters or fewer")

    pickup = _resolve_address(
        customer,
        address_id=pickup_address_id,
        address_data=pickup_address,
        default_label="Pickup",
    )
    dropoff = _resolve_address(
        customer,
        address_id=dropoff_address_id,
        address_data=dropoff_address,
        default_label="Drop-off",
    )

    if pickup.id == dropoff.id:
        raise ValueError("Pickup and drop-off must be different addresses")

    fee, fee_pending = estimate_logistics_fee(pickup.parish, dropoff.parish)
    if vehicle == "truck":
        fee_pending = True
        fee = None

    job = LogisticsJob(
        reference=_next_reference(),
        customer_id=customer.id,
        pickup_address_id=pickup.id,
        dropoff_address_id=dropoff.id,
        item_description=description,
        vehicle_type=vehicle,
        delivery_speed=speed,
        payment_method=pay_method,
        weight_lbs=None,
        notes=note_text,
        status="pending",
        quoted_fee_jmd=None if fee_pending else fee,
        fee_pending_quote=fee_pending,
        requested_at=datetime.utcnow(),
    )
    db.session.add(job)
    db.session.commit()
    return job


def update_customer_logistics_notes(job: LogisticsJob, customer: User, notes: str) -> LogisticsJob:
    if job.customer_id != customer.id:
        raise ValueError("Local delivery request not found")
    if job.status not in LOGISTICS_JOB_OPEN_STATUSES:
        raise ValueError("Notes can only be updated on active deliveries")

    note_text = (notes or "").strip()
    if not note_text:
        raise ValueError("Note cannot be empty")
    if len(note_text) > 500:
        raise ValueError("Note must be 500 characters or fewer")

    job.notes = note_text
    db.session.commit()
    return job


def confirm_logistics_driver(
    job: LogisticsJob,
    staff_user: User,
    *,
    driver_name: str,
    driver_contact_number: str,
) -> LogisticsJob:
    if job.status != "pending":
        raise ValueError("Driver can only be confirmed on jobs searching for a driver")

    name = (driver_name or "").strip()
    if not name:
        raise ValueError("Driver name is required")

    from app.services.auth_service import normalize_phone

    try:
        contact = normalize_phone(driver_contact_number)
    except ValueError as exc:
        raise ValueError(str(exc)) from exc

    job.driver_name = name
    job.driver_contact_number = contact
    job.driver_confirmed_at = datetime.utcnow()
    job.driver_confirmed_by_id = staff_user.id
    db.session.commit()
    return job


def _get_active_clerk(clerk_id) -> User:
    import uuid

    try:
        cid = uuid.UUID(str(clerk_id))
    except ValueError as exc:
        raise ValueError("Select a valid clerk") from exc

    clerk = User.query.filter_by(id=cid, role="clerk").first()
    if not clerk or clerk.is_active is False:
        raise ValueError("Clerk not found or inactive")
    if not (clerk.contact_number or "").strip():
        raise ValueError("Clerk must have a contact number before assignment")
    return clerk


def assign_logistics_clerk(job: LogisticsJob, admin_user: User, *, clerk_id) -> LogisticsJob:
    if job.status != "pending":
        raise ValueError("Only pending local delivery requests can be assigned")

    clerk = _get_active_clerk(clerk_id)

    from app.services.auth_service import normalize_phone

    try:
        contact = normalize_phone(clerk.contact_number)
    except ValueError as exc:
        raise ValueError(f"Clerk contact number is invalid: {exc}") from exc

    now = datetime.utcnow()
    job.assigned_clerk_id = clerk.id
    job.assigned_at = now
    job.assigned_by_id = admin_user.id
    job.driver_name = clerk.full_name
    job.driver_contact_number = contact
    job.driver_confirmed_at = now
    job.driver_confirmed_by_id = admin_user.id
    db.session.commit()
    refreshed = get_logistics_job(job.id)
    return refreshed or job


def reject_logistics_job(
    job: LogisticsJob,
    admin_user: User,
    *,
    reason: str | None = None,
) -> LogisticsJob:
    if job.status != "pending":
        raise ValueError("Only pending local delivery requests can be rejected")

    reason_text = (reason or "").strip() or None
    if reason_text and len(reason_text) > 500:
        raise ValueError("Rejection reason must be 500 characters or fewer")

    job.status = "rejected"
    job.rejected_at = datetime.utcnow()
    job.rejected_by_id = admin_user.id
    job.rejection_reason = reason_text
    db.session.commit()
    return job


def cancel_logistics_job(job: LogisticsJob, *, by_customer: bool = True) -> LogisticsJob:
    if by_customer and job.status != "pending":
        raise ValueError("Only pending jobs can be cancelled")
    if not by_customer and job.status not in LOGISTICS_JOB_OPEN_STATUSES:
        raise ValueError("Only open jobs can be cancelled")

    job.status = "cancelled"
    job.cancelled_at = datetime.utcnow()
    db.session.commit()
    return job


def mark_logistics_job_picked_up(job: LogisticsJob, staff_user: User) -> LogisticsJob:
    if job.status != "pending":
        raise ValueError("Only jobs searching for a driver can be marked picked up")

    job.status = "picked_up"
    job.picked_up_at = datetime.utcnow()
    job.picked_up_by_id = staff_user.id
    db.session.commit()
    return job


def mark_logistics_job_in_transit(job: LogisticsJob, staff_user: User) -> LogisticsJob:
    if job.status != "picked_up":
        raise ValueError("Only picked-up jobs can be marked in transit")

    job.status = "in_transit"
    job.in_transit_at = datetime.utcnow()
    job.in_transit_by_id = staff_user.id
    db.session.commit()
    return job


def complete_logistics_job(job: LogisticsJob, staff_user: User) -> LogisticsJob:
    if job.status != "in_transit":
        raise ValueError("Only in-transit jobs can be completed")

    job.status = "completed"
    job.completed_at = datetime.utcnow()
    job.completed_by_id = staff_user.id
    db.session.commit()
    return job


def confirm_customer_logistics_receipt(job: LogisticsJob, customer: User) -> LogisticsJob:
    if job.customer_id != customer.id:
        raise ValueError("Local delivery request not found")
    if job.status != "completed":
        raise ValueError("Receipt can only be confirmed after delivery")
    if job.customer_receipt_confirmed_at:
        raise ValueError("Receipt already confirmed")

    job.customer_receipt_confirmed_at = datetime.utcnow()
    db.session.commit()
    return job

File: server/src/services/logisticsJobService.ts
import { and, asc, count, desc, eq, inArray, like, notInArray } from 'drizzle-orm'
import { validate as validateUuid } from 'uuid'
import {
  LOGISTICS_DELIVERY_SPEEDS,
  LOGISTICS_ENABLED_PAYMENT_METHODS,
  LOGISTICS_IN_HOUSE_FEE_JMD,
  LOGISTICS_IN_HOUSE_PARISHES,
  LOGISTICS_ISLAND_FEE_JMD,
  LOGISTICS_ITEM_CATEGORIES,
  LOGISTICS_ITEM_CATEGORY_LABELS,
  LOGISTICS_JOB_OPEN_STATUSES,
  LOGISTICS_PAYMENT_METHODS,
  LOGISTICS_VEHICLE_TYPES,
} from '../constants.js'
import { db } from '../db/index.js'
import {
  deliveryAddresses,
  logisticsJobs,
  users,
  type UserRow,
} from '../db/schema/index.js'
import { logisticsJobToDict } from '../lib/serializers/logisticsJob.js'
import { userFullName } from '../lib/user.js'
import { normalizePhone } from './authService.js'
import { createDeliveryAddress, getDeliveryAddress } from './deliveryAddressService.js'

type LogisticsJobRow = typeof logisticsJobs.$inferSelect
type DeliveryAddressRow = typeof deliveryAddresses.$inferSelect

const OPEN_STATUSES = [...LOGISTICS_JOB_OPEN_STATUSES]

export async function serializeJob(job: LogisticsJobRow): Promise<Record<string, unknown>> {
  const userIds = [
    job.customerId,
    job.inProgressById,
    job.pickedUpById,
    job.inTransitById,
    job.driverConfirmedById,
    job.completedById,
    job.assignedClerkId,
    job.assignedById,
    job.rejectedById,
  ].filter((id): id is string => Boolean(id))

  const relatedUsers =
    userIds.length > 0
      ? await db.select().from(users).where(inArray(users.id, [...new Set(userIds)]))
      : []
  const userById = new Map(relatedUsers.map((u) => [u.id, u]))

  const addressIds = [job.pickupAddressId, job.dropoffAddressId]
  const relatedAddresses = await db
    .select()
    .from(deliveryAddresses)
    .where(inArray(deliveryAddresses.id, addressIds))
  const addressById = new Map(relatedAddresses.map((a) => [a.id, a]))

  return logisticsJobToDict(job, {
    customer: userById.get(job.customerId) ?? null,
    pickupAddress: addressById.get(job.pickupAddressId) ?? null,
    dropoffAddress: addressById.get(job.dropoffAddressId) ?? null,
    assignedClerk: job.assignedClerkId ? userById.get(job.assignedClerkId) ?? null : null,
    assignedBy: job.assignedById ? userById.get(job.assignedById) ?? null : null,
    rejectedBy: job.rejectedById ? userById.get(job.rejectedById) ?? null : null,
    inProgressBy: job.inProgressById ? userById.get(job.inProgressById) ?? null : null,
    pickedUpBy: job.pickedUpById ? userById.get(job.pickedUpById) ?? null : null,
    inTransitBy: job.inTransitById ? userById.get(job.inTransitById) ?? null : null,
    completedBy: job.completedById ? userById.get(job.completedById) ?? null : null,
  })
}

async function nextReference(): Promise<string> {
  const prefix = 'LD'
  const [latest] = await db
    .select()
    .from(logisticsJobs)
    .where(like(logisticsJobs.reference, `${prefix}-%`))
    .orderBy(desc(logisticsJobs.requestedAt))
    .limit(1)

  let seq: number
  if (latest?.reference.startsWith(`${prefix}-`)) {
    try {
      seq = Number.parseInt(latest.reference.split('-', 2)[1] ?? '', 10) + 1
      if (!Number.isFinite(seq)) throw new Error('bad seq')
    } catch {
      const [total] = await db.select({ value: count() }).from(logisticsJobs)
      seq = (total?.value ?? 0) + 1
    }
  } else {
    const [total] = await db.select({ value: count() }).from(logisticsJobs)
    seq = (total?.value ?? 0) + 1
  }
  return `${prefix}-${String(seq).padStart(5, '0')}`
}

async function resolveAddress(
  customer: UserRow,
  opts: {
    addressId?: string | null
    addressData?: Record<string, unknown> | null
    defaultLabel: string
  },
): Promise<DeliveryAddressRow> {
  if (opts.addressId) {
    const address = await getDeliveryAddress(customer, String(opts.addressId))
    if (!address) throw new Error(`${opts.defaultLabel} address not found`)
    return address
  }

  if (opts.addressData && typeof opts.addressData === 'object') {
    const payload = { ...opts.addressData }
    if (!String(payload.label ?? '').trim()) {
      payload.label = opts.defaultLabel
    }
    return createDeliveryAddress(customer, payload)
  }

  throw new Error(`${opts.defaultLabel} address is required`)
}

export function estimateLogisticsFee(
  pickupParish: string,
  dropoffParish: string,
): { fee: string | null; feePending: boolean } {
  const pickupInHouse = (LOGISTICS_IN_HOUSE_PARISHES as readonly string[]).includes(pickupParish)
  const dropoffInHouse = (LOGISTICS_IN_HOUSE_PARISHES as readonly string[]).includes(dropoffParish)
  if (pickupInHouse && dropoffInHouse) {
    return { fee: LOGISTICS_IN_HOUSE_FEE_JMD, feePending: false }
  }
  if (pickupInHouse || dropoffInHouse) {
    return { fee: LOGISTICS_ISLAND_FEE_JMD, feePending: false }
  }
  return { fee: null, feePending: true }
}

export async function listCustomerLogisticsJobs(
  customer: UserRow,
  limit = 50,
): Promise<LogisticsJobRow[]> {
  return db
    .select()
    .from(logisticsJobs)
    .where(eq(logisticsJobs.customerId, customer.id))
    .orderBy(desc(logisticsJobs.requestedAt))
    .limit(limit)
}

export async function listPendingLogisticsJobs(limit = 100): Promise<LogisticsJobRow[]> {
  return db
    .select()
    .from(logisticsJobs)
    .where(eq(logisticsJobs.status, 'pending'))
    .orderBy(asc(logisticsJobs.requestedAt))
    .limit(limit)
}

export async function listOpenLogisticsJobs(limit = 100): Promise<LogisticsJobRow[]> {
  return db
    .select()
    .from(logisticsJobs)
    .where(inArray(logisticsJobs.status, OPEN_STATUSES))
    .orderBy(asc(logisticsJobs.requestedAt))
    .limit(limit)
}

export async function listAllLogisticsJobs(limit = 200): Promise<LogisticsJobRow[]> {
  return db
    .select()
    .from(logisticsJobs)
    .orderBy(desc(logisticsJobs.requestedAt))
    .limit(limit)
}

export async function listLogisticsJobHistory(limit = 200): Promise<LogisticsJobRow[]> {
  return db
    .select()
    .from(logisticsJobs)
    .where(notInArray(logisticsJobs.status, OPEN_STATUSES))
    .orderBy(desc(logisticsJobs.requestedAt))
    .limit(limit)
}

export async function listLogisticsJobsByStatus(
  status: string,
  limit = 200,
): Promise<LogisticsJobRow[]> {
  return db
    .select()
    .from(logisticsJobs)
    .where(eq(logisticsJobs.status, status))
    .orderBy(desc(logisticsJobs.requestedAt))
    .limit(limit)
}

export async function countOpenLogisticsJobs(): Promise<number> {
  const [row] = await db
    .select({ value: count() })
    .from(logisticsJobs)
    .where(inArray(logisticsJobs.status, OPEN_STATUSES))
  return row?.value ?? 0
}

export async function countClerkOpenLogisticsJobs(clerk: UserRow): Promise<number> {
  const [row] = await db
    .select({ value: count() })
    .from(logisticsJobs)
    .where(
      and(
        eq(logisticsJobs.assignedClerkId, clerk.id),
        inArray(logisticsJobs.status, OPEN_STATUSES),
      ),
    )
  return row?.value ?? 0
}

export async function listClerkLogisticsJobs(
  clerk: UserRow,
  status = 'active',
  limit = 100,
): Promise<LogisticsJobRow[]> {
  const conditions = [eq(logisticsJobs.assignedClerkId, clerk.id)]

  if (status === 'pending') {
    conditions.push(eq(logisticsJobs.status, 'pending'))
  } else if (status === 'active') {
    conditions.push(inArray(logisticsJobs.status, OPEN_STATUSES))
  } else if (status === 'history') {
    conditions.push(notInArray(logisticsJobs.status, OPEN_STATUSES))
  } else if (status !== 'all') {
    conditions.push(eq(logisticsJobs.status, status))
  }

  return db
    .select()
    .from(logisticsJobs)
    .where(and(...conditions))
    .orderBy(asc(logisticsJobs.requestedAt))
    .limit(limit)
}

export async function countPendingLogisticsJobs(): Promise<number> {
  const [row] = await db
    .select({ value: count() })
    .from(logisticsJobs)
    .where(eq(logisticsJobs.status, 'pending'))
  return row?.value ?? 0
}

export async function getLogisticsJob(jobId: string): Promise<LogisticsJobRow | null> {
  if (!validateUuid(String(jobId))) return null
  const [row] = await db
    .select()
    .from(logisticsJobs)
    .where(eq(logisticsJobs.id, String(jobId)))
    .limit(1)
  return row ?? null
}

export function resolveItemDescription(opts: {
  itemCategory?: string | null
  itemOtherDetail?: string | null
  itemDescription?: string | null
}): string {
  if (opts.itemCategory) {
    const category = opts.itemCategory.trim().toLowerCase()
    if (!(LOGISTICS_ITEM_CATEGORIES as readonly string[]).includes(category)) {
      throw new Error('Select a valid item category')
    }
    if (category === 'other') {
      const detail = (opts.itemOtherDetail || '').trim()
      if (!detail) throw new Error('Describe the item when selecting Other')
      if (detail.length > 480) {
        throw new Error('Item description must be 500 characters or fewer')
      }
      return `Other — ${detail}`
    }
    return LOGISTICS_ITEM_CATEGORY_LABELS[category]!
  }

  const description = (opts.itemDescription || '').trim()
  if (!description) throw new Error('Item category is required')
  if (description.length > 500) {
    throw new Error('Item description must be 500 characters or fewer')
  }
  return description
}

export function resolveVehicleType(vehicleType: string | null | undefined): string {
  const value = (vehicleType || '').trim().toLowerCase()
  if (!(LOGISTICS_VEHICLE_TYPES as readonly string[]).includes(value)) {
    throw new Error('Select a vehicle size for this delivery')
  }
  return value
}

export function resolveDeliverySpeed(deliverySpeed: string | null | undefined): string {
  const value = (deliverySpeed || 'immediate').trim().toLowerCase()
  if (!(LOGISTICS_DELIVERY_SPEEDS as readonly string[]).includes(value)) {
    throw new Error('Select when you need this delivery')
  }
  return value
}

export function resolvePaymentMethod(paymentMethod: string | null | undefined): string {
  const value = (paymentMethod || 'cash').trim().toLowerCase()
  if (!(LOGISTICS_PAYMENT_METHODS as readonly string[]).includes(value)) {
    throw new Error('Select a payment method')
  }
  if (!(LOGISTICS_ENABLED_PAYMENT_METHODS as readonly string[]).includes(value)) {
    throw new Error('Online payment is not available yet — please choose cash')
  }
  return value
}

export async function createLogisticsJob(
  customer: UserRow,
  opts: {
    pickupAddressId?: string | null
    pickupAddress?: Record<string, unknown> | null
    dropoffAddressId?: string | null
    dropoffAddress?: Record<string, unknown> | null
    itemCategory?: string | null
    itemOtherDetail?: string | null
    itemDescription?: string | null
    vehicleType?: string | null
    deliverySpeed?: string | null
    paymentMethod?: string | null
    weightLbs?: number | string | null
    notes?: string | null
  },
): Promise<LogisticsJobRow> {
  const description = resolveItemDescription({
    itemCategory: opts.itemCategory,
    itemOtherDetail: opts.itemOtherDetail,
    itemDescription: opts.itemDescription,
  })
  const vehicle = resolveVehicleType(opts.vehicleType)
  const speed = resolveDeliverySpeed(opts.deliverySpeed)
  const payMethod = resolvePaymentMethod(opts.paymentMethod)

  const noteText = (opts.notes || '').trim() || null
  if (noteText && noteText.length > 500) {
    throw new Error('notes must be 500 characters or fewer')
  }

  const pickup = await resolveAddress(customer, {
    addressId: opts.pickupAddressId,
    addressData: opts.pickupAddress,
    defaultLabel: 'Pickup',
  })
  const dropoff = await resolveAddress(customer, {
    addressId: opts.dropoffAddressId,
    addressData: opts.dropoffAddress,
    defaultLabel: 'Drop-off',
  })

  if (pickup.id === dropoff.id) {
    throw new Error('Pickup and drop-off must be different addresses')
  }

  let { fee, feePending } = estimateLogisticsFee(pickup.parish, dropoff.parish)
  if (vehicle === 'truck') {
    feePending = true
    fee = null
  }

  const [job] = await db
    .insert(logisticsJobs)
    .values({
      reference: await nextReference(),
      customerId: customer.id,
      pickupAddressId: pickup.id,
      dropoffAddressId: dropoff.id,
      itemDescription: description,
      vehicleType: vehicle,
      deliverySpeed: speed,
      paymentMethod: payMethod,
      weightLbs: null,
      notes: noteText,
      status: 'pending',
      quotedFeeJmd: feePending ? null : fee,
      feePendingQuote: feePending,
      requestedAt: new Date(),
    })
    .returning()

  return job!
}

export async function updateCustomerLogisticsNotes(
  job: LogisticsJobRow,
  customer: UserRow,
  notes: string,
): Promise<LogisticsJobRow> {
  if (job.customerId !== customer.id) {
    throw new Error('Local delivery request not found')
  }
  if (!(OPEN_STATUSES as readonly string[]).includes(job.status)) {
    throw new Error('Notes can only be updated on active deliveries')
  }

  const noteText = (notes || '').trim()
  if (!noteText) throw new Error('Note cannot be empty')
  if (noteText.length > 500) throw new Error('Note must be 500 characters or fewer')

  const [updated] = await db
    .update(logisticsJobs)
    .set({ notes: noteText })
    .where(eq(logisticsJobs.id, job.id))
    .returning()
  return updated!
}

export async function confirmLogisticsDriver(
  job: LogisticsJobRow,
  staffUser: UserRow,
  opts: { driverName: string; driverContactNumber: string },
): Promise<LogisticsJobRow> {
  if (job.status !== 'pending') {
    throw new Error('Driver can only be confirmed on jobs searching for a driver')
  }

  const name = (opts.driverName || '').trim()
  if (!name) throw new Error('Driver name is required')

  let contact: string
  try {
    contact = normalizePhone(opts.driverContactNumber)
  } catch (err) {
    throw new Error(err instanceof Error ? err.message : String(err))
  }

  const [updated] = await db
    .update(logisticsJobs)
    .set({
      driverName: name,
      driverContactNumber: contact,
      driverConfirmedAt: new Date(),
      driverConfirmedById: staffUser.id,
    })
    .where(eq(logisticsJobs.id, job.id))
    .returning()
  return updated!
}

async function getActiveClerk(clerkId: unknown): Promise<UserRow> {
  if (!validateUuid(String(clerkId))) {
    throw new Error('Select a valid clerk')
  }

  const [clerk] = await db
    .select()
    .from(users)
    .where(and(eq(users.id, String(clerkId)), eq(users.role, 'clerk')))
    .limit(1)

  if (!clerk || clerk.isActive === false) {
    throw new Error('Clerk not found or inactive')
  }
  if (!(clerk.contactNumber || '').trim()) {
    throw new Error('Clerk must have a contact number before assignment')
  }
  return clerk
}

export async function assignLogisticsClerk(
  job: LogisticsJobRow,
  adminUser: UserRow,
  opts: { clerkId: unknown },
): Promise<LogisticsJobRow> {
  if (job.status !== 'pending') {
    throw new Error('Only pending local delivery requests can be assigned')
  }

  const clerk = await getActiveClerk(opts.clerkId)

  let contact: string
  try {
    contact = normalizePhone(clerk.contactNumber!)
  } catch (err) {
    throw new Error(
      `Clerk contact number is invalid: ${err instanceof Error ? err.message : String(err)}`,
    )
  }

  const now = new Date()
  await db
    .update(logisticsJobs)
    .set({
      assignedClerkId: clerk.id,
      assignedAt: now,
      assignedById: adminUser.id,
      driverName: userFullName(clerk),
      driverContactNumber: contact,
      driverConfirmedAt: now,
      driverConfirmedById: adminUser.id,
    })
    .where(eq(logisticsJobs.id, job.id))

  return (await getLogisticsJob(job.id)) ?? job
}

export async function rejectLogisticsJob(
  job: LogisticsJobRow,
  adminUser: UserRow,
  opts: { reason?: string | null } = {},
): Promise<LogisticsJobRow> {
  if (job.status !== 'pending') {
    throw new Error('Only pending local delivery requests can be rejected')
  }

  const reasonText = (opts.reason || '').trim() || null
  if (reasonText && reasonText.length > 500) {
    throw new Error('Rejection reason must be 500 characters or fewer')
  }

  const [updated] = await db
    .update(logisticsJobs)
    .set({
      status: 'rejected',
      rejectedAt: new Date(),
      rejectedById: adminUser.id,
      rejectionReason: reasonText,
    })
    .where(eq(logisticsJobs.id, job.id))
    .returning()
  return updated!
}

export async function cancelLogisticsJob(
  job: LogisticsJobRow,
  opts: { byCustomer?: boolean } = {},
): Promise<LogisticsJobRow> {
  const byCustomer = opts.byCustomer !== false
  if (byCustomer && job.status !== 'pending') {
    throw new Error('Only pending jobs can be cancelled')
  }
  if (!byCustomer && !(OPEN_STATUSES as readonly string[]).includes(job.status)) {
    throw new Error('Only open jobs can be cancelled')
  }

  const [updated] = await db
    .update(logisticsJobs)
    .set({
      status: 'cancelled',
      cancelledAt: new Date(),
    })
    .where(eq(logisticsJobs.id, job.id))
    .returning()
  return updated!
}

export async function markLogisticsJobPickedUp(
  job: LogisticsJobRow,
  staffUser: UserRow,
): Promise<LogisticsJobRow> {
  if (job.status !== 'pending') {
    throw new Error('Only jobs searching for a driver can be marked picked up')
  }

  const [updated] = await db
    .update(logisticsJobs)
    .set({
      status: 'picked_up',
      pickedUpAt: new Date(),
      pickedUpById: staffUser.id,
    })
    .where(eq(logisticsJobs.id, job.id))
    .returning()
  return updated!
}

export async function markLogisticsJobInTransit(
  job: LogisticsJobRow,
  staffUser: UserRow,
): Promise<LogisticsJobRow> {
  if (job.status !== 'picked_up') {
    throw new Error('Only picked-up jobs can be marked in transit')
  }

  const [updated] = await db
    .update(logisticsJobs)
    .set({
      status: 'in_transit',
      inTransitAt: new Date(),
      inTransitById: staffUser.id,
    })
    .where(eq(logisticsJobs.id, job.id))
    .returning()
  return updated!
}

export async function completeLogisticsJob(
  job: LogisticsJobRow,
  staffUser: UserRow,
): Promise<LogisticsJobRow> {
  if (job.status !== 'in_transit') {
    throw new Error('Only in-transit jobs can be completed')
  }

  const [updated] = await db
    .update(logisticsJobs)
    .set({
      status: 'completed',
      completedAt: new Date(),
      completedById: staffUser.id,
    })
    .where(eq(logisticsJobs.id, job.id))
    .returning()
  return updated!
}

export async function confirmCustomerLogisticsReceipt(
  job: LogisticsJobRow,
  customer: UserRow,
): Promise<LogisticsJobRow> {
  if (job.customerId !== customer.id) {
    throw new Error('Local delivery request not found')
  }
  if (job.status !== 'completed') {
    throw new Error('Receipt can only be confirmed after delivery')
  }
  if (job.customerReceiptConfirmedAt) {
    throw new Error('Receipt already confirmed')
  }

  const [updated] = await db
    .update(logisticsJobs)
    .set({ customerReceiptConfirmedAt: new Date() })
    .where(eq(logisticsJobs.id, job.id))
    .returning()
  return updated!
}
